cgi/product_multilingual.pl

```diff
@@ -1101,42 +1104,54 @@ ($product_ref, $field, $language, $request_ref)
 
 	my $hidden_inputs = '';
 
-	#<p class="note">&rarr; $Lang{nutrition_data_table_note}{$lc}</p>
+	# Nutrition data
+	#
+	# With the new nutrition data model introduced in 2025, the product edit form on the web
+	# will only allow to edit input sets for a single source:
+	# - the "packaging" source on the public platform
+	# - the "manufacturer" source for most organizations on the producers platform
+	# - the "database-[name]" source for organizations with an id org-database-[name] (e.g. org-database-usda)
+	#
+	# 6 input sets will be writable for the selected source:
+	# - as sold per 100g
+	# - as sold per 100ml
+	# - as sold per serving
+	# - prepared per 100g
+	# - prepared per 100ml
+	# - prepared per serving
+	#
+	# The API v2 will be extended to allow editing those input sets
 
 	# We first go through all nutrients, so that we can see the ones for which we have data
 	# as we will check the nutrition_data checkbox if we have data for at least one nutrient
 
-	defined $product_ref->{nutriments} or $product_ref->{nutriments} = {};
+	my @preparations = get_preparations_for_product_type($product_ref->{product_type});
+	my @pers = get_pers_for_product_type($product_ref->{product_type});
 
-	my @unknown_nutriments = ();
-	my %seen_unknown_nutriments = ();
-	foreach my $nid (keys %{$product_ref->{nutriments}}) {
+	my $input_sets_hash_ref = get_nutrition_input_sets_in_a_hash($product_ref);
 
-		next if (($nid =~ /_/) and ($nid !~ /_prepared$/));
+	# Go through all nutrients and all input sets
+	# We will display a table with one row per nutrient and one column per input set
 
-		$nid =~ s/_prepared$//;
+	# If we have a value for at least one nutrient in one input set, we will display the corresponding column
+	my %input_sets = ();
 
-		$log->trace("detect unknown nutriment", {nid => $nid}) if $log->is_trace();
+	# If we have a value for at least one input set for one nutrient, we will display the corresponding line
+	# @nutrients will contain the list of nutrients to display with their values and units for each input set
+	my @nutrients = ();
 
-		if (    (not exists_taxonomy_tag("nutrients", "zz:$nid"))
-			and (defined $product_ref->{nutriments}{$nid . "_label"})
-			and (not defined $seen_unknown_nutriments{$nid}))
-		{
-			push @unknown_nutriments, $nid;
-			$log->debug("unknown nutriment detected", {nid => $nid}) if $log->is_debug();
-		}
-	}
-
-	# Go through all nutrients
+	$template_data_ref_display->{source} = $source;
+	$template_data_ref_display->{preparations} = \@preparations;
+	$template_data_ref_display->{pers} = \@pers;
+	$template_data_ref_display->{nutrients} = \@nutrients;
+	$template_data_ref_display->{input_sets} = \%input_sets;
 
-	my %nutrition_data_exists = ();
-	my @nutriments;
-	foreach my $nutriment (@{$nutriments_tables{$nutriment_table}}, @unknown_nutriments, 'new_0', 'new_1') {
+	foreach my $nutrient (@{$nutriments_tables{$nutriment_table}}) {
```

Comment:
It would be worth explaining what structure we are building. Because it's hard to understand the code.
It's optional of course.